### PR TITLE
Enable instant execution test coverage CI pipeline

### PR DESCRIPTION
**Pull request by eskatos**
This PR enables the instant execution test coverage CI pipeline introduced in https://github.com/gradle/gradle/pull/11211

It enables the `:instantIntegTest` tasks in all sub projects except `:instantExecution`.

It limits the number of instant execution problems to fail the build to `1` in order to prevent OOMEs caused by the test fixtures holding multiple references to the whole output that gets really large because of huge stacktraces.

It also introduces a flag to make instant execution run quietly in order to not disturb assertions on the full build output.

It changes the executer from in-process to daemon because using an in-process executer caused lots of troubles like OOMEs, hanging builds and other weirdness, see the related commit for some more details.

It removes the now unneeded `InstantExecutionRunner`.

It introduces `@FailsWithInstantExecution` and `@IgnoreWithInstantExecution` annotations and their respective Spock extensions and JUnit rules and annotates integration tests accordingly, hence the large number of commits and large changeset.

We should try to merge this quickly given the high probability of conflicts. If we really want to move back to the in-process executer for faster feedback let's discuss this separately.

### CODE REVIEW

File: subprojects/testing-base/src/testFixtures/groovy/org/gradle/testing/AbstractTestFrameworkIntegrationTest.groovy [hunk 1]
```diff
@@ -96,6 +98,7 @@ abstract class AbstractTestFrameworkIntegrationTest extends AbstractIntegrationS
         succeeds "verifyTestResultConventions"
     }
 
+    @IgnoreWithInstantExecution
```
Comment (adammurdoch) [inline subprojects/testing-base/src/testFixtures/groovy/org/gradle/testing/AbstractTestFrameworkIntegrationTest.groovy]: Why do we want to ignore all these testing tests with instant execution?

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiIntegrationTest.groovy [hunk 2]
```diff
@@ -70,6 +71,7 @@ class ToolingApiIntegrationTest extends AbstractIntegrationSpec {
         !stdOut.toString().contains("BUILD SUCCESSFUL")
     }
 
+    @IgnoreWithInstantExecution
```
Comment (adammurdoch) [inline subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiIntegrationTest.groovy]: Why ignore?

File: subprojects/version-control/src/integTest/groovy/org/gradle/vcs/internal/AbstractSourceDependencyMultiprojectIntegrationTest.groovy [hunk 3]
```diff
@@ -71,6 +72,7 @@ abstract class AbstractSourceDependencyMultiprojectIntegrationTest extends Abstr
         """
     }
 
+    @IgnoreWithInstantExecution
```
Comment (adammurdoch) [inline subprojects/version-control/src/integTest/groovy/org/gradle/vcs/internal/AbstractSourceDependencyMultiprojectIntegrationTest.groovy]: Why ignore?

File: subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorIntegrationTest.groovy [hunk 4]
```diff
@@ -39,6 +40,7 @@ class WorkerExecutorIntegrationTest extends AbstractWorkerExecutorIntegrationTes
 
     def buildOperations = new BuildOperationsFixture(executer, temporaryFolder)
 
+    @IgnoreWithInstantExecution
```
Comment (adammurdoch) [inline subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorIntegrationTest.groovy]: why ignore?

File: subprojects/core/src/integTest/groovy/org/gradle/api/services/BuildServiceIntegrationTest.groovy [hunk 5]
```diff
@@ -17,12 +17,8 @@
 package org.gradle.api.services
 
 import org.gradle.integtests.fixtures.AbstractIntegrationSpec
-import org.gradle.integtests.fixtures.InstantExecutionRunner
-import org.gradle.integtests.fixtures.RequiredFeature
-import org.gradle.integtests.fixtures.RequiredFeatures
-import org.junit.runner.RunWith
+import org.gradle.integtests.fixtures.FailsWithInstantExecution
 
-@RunWith(InstantExecutionRunner)
```
[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/core/src/integTest/groovy/org/gradle/api/services/BuildServiceIntegrationTest.groovy]: Good point about the runner and debugging, I'll reinstate it so we can use it on demand for debugging.

**Comment by eskatos** [comment]
@adammurdoch, most of the tests annotated with `@IgnoreWithInstantExecution` fall into one of two buckets:
- unrolled tests where some cases pass and some fail ;
- tests in abstract classes where some descendants pass and some fail - we can't override tests in Spock (known limitation) to annotate descendants accordingly ;

If Spock allowed to annotate unrolled cases separately or to override features we could annotate each test more precisely. Ignoring them is the best strategy I came up with, any idea on how this could be handled differently?

Then there are some other tests I might have ignored while using the in-process executer that caused all kind of weirdnesses like timeouts due to some concurrency fixtures not detecting some event. I'm going to disable the ignoring, run the whole pipeline again to find them and remove the spurious ignores.

**Comment by eskatos** [comment]
After further analysis there's another bucket of ignored tests that failed after the Spock extensions had a chance to do something because they use JUnit rules such as `HttpServer`. While investigating how to fix that I stumbled upon a bug in Spock handling of JUnit rules, they aren't nested in the correct order ... see the last commits above. Moving back to _In Progress_

**Comment by adammurdoch** [comment]
I think for a method for which some iterations fail and some pass, I would still mark these as 'fails' but in a way that disables the check that verifies that the test does actually fail. For me, 'ignore' does not communicate the intent very well and is easy to accidentally use instead of 'fails', in which case we lose the check that will let us know when the test starts to work.

Another variation would be to move the check somewhere outside of the test rule, so we can check that one or more of the iterations did fail.

File: subprojects/core/src/integTest/groovy/org/gradle/api/services/BuildServiceIntegrationTest.groovy [hunk 6]
```diff
@@ -124,9 +120,7 @@ class BuildServiceIntegrationTest extends AbstractIntegrationSpec {
         outputContains("service: closed with value 12")
     }
 
-    @RequiredFeatures(
-        [@RequiredFeature(feature = "org.gradle.unsafe.instant-execution", value="false")]
-    )
+    @FailsWithInstantExecution
```
[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/core/src/integTest/groovy/org/gradle/api/services/BuildServiceIntegrationTest.groovy]: I went with ignoring specifying a reason of _unsupported_.

**Comment by eskatos** [comment]
I reworked the spock extensions, junit rules, how they interact and reviewed all ignored tests and things are in a much better shape now.

`@FailsWithInstantExecution` now supports unrolled tests where not all iterations fail.

`@IgnoreWithInstantExecution` now requires a _reason_ which can be _unsupported_, _fails in subclass_, _flaky_, _long timeout_ or _requires investigation_. See the enum for detailed documentation on the intent of each possible reason.

From ~300 spock features annotated to be ignored by `:instantIntegTest` we're down to 67.

0 require investigation at this stage, but I kept that reason so that other people can tag tests with it when encountering some not understood problem.
1 is unsupported in `BuildServiceIntegrationTest`, we want to import org.gradle.integtests.fixtures.FailsWithInstantExecution
remove it not fix it.
60 declared in abstract specs fail in some subclass, this is due to the spock limitation that prevent to  override features and annotate them in subclasses, we can't annotate them as failing.
5 are flaky only with instant execution, we can't annotate them as failing.
1 class is ignored because it takes ~30min to timeout and would significantly slow down the ci feedback.

In any case, we should aim at removing all those annotations progressively. I think this is in an acceptable state to be merged and followed up on.

@adammurdoch please take another look

File: subprojects/core/src/integTest/groovy/org/gradle/api/services/BuildServiceIntegrationTest.groovy [hunk 7]
```diff
@@ -17,12 +17,8 @@
 package org.gradle.api.services
 
 import org.gradle.integtests.fixtures.AbstractIntegrationSpec
-import org.gradle.integtests.fixtures.InstantExecutionRunner
-import org.gradle.integtests.fixtures.RequiredFeature
-import org.gradle.integtests.fixtures.RequiredFeatures
-import org.junit.runner.RunWith
+import org.gradle.integtests.fixtures.IgnoreWithInstantExecution
 
-@RunWith(InstantExecutionRunner)
```
Comment (adammurdoch) [inline subprojects/core/src/integTest/groovy/org/gradle/api/services/BuildServiceIntegrationTest.groovy]: Let's keep the runner attached to this test while we're working on the feature that it covers and while the instant execution suite does not use the in-process executor.

Comment (eskatos) [reply subprojects/core/src/integTest/groovy/org/gradle/api/services/BuildServiceIntegrationTest.groovy]: Done

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/IgnoreWithInstantExecution.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        /**
+         * Use this reason on tests that take a long time to fail, slowing down the CI feedback.
+         * Use sparingly, only in dramatic cases.
+         */
+        LONG_TIMEOUT,
+
+        /**
+         * Use this reason if you don't know what is happening and want to postpone the investigation.
+         */
+        REQUIRES_INVESTIGATION
```
[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/IgnoreWithInstantExecution.java]: Good points. I went ahead and made the changes in this PR.

**Comment by eskatos** [comment]
There's now `@UnsupportedWithInstantExecution` to convey that the annotated test will never be supported with instant execution.

And `@ToBeFixedForInstantExecution` that has an optional reason for skipping running the test with instant execution. If no reason is provided then the test is executed and expected to fail.

**Pull request merged**